src/utils/merge-nodes-with-matching-flavors.ts

```diff
@@ -0,0 +1,166 @@
+            }
+            return groups;
+        },
+        {
+            [importFlavorRegular]: [] as ImportDeclaration[],
+            [importFlavorType]: [] as ImportDeclaration[],
+        },
+    );
+}
+
+function selectNodeImportSource(node: ImportDeclaration) {
+    return node.source.value;
+}
+
+function nodeIsImportNamespaceSpecifier(
+    node: ImportSpecifier | ImportDefaultSpecifier | ImportNamespaceSpecifier,
+): node is ImportNamespaceSpecifier {
+    return node.type === 'ImportNamespaceSpecifier';
+}
+function nodeIsImportDefaultSpecifier(
+    node: ImportSpecifier | ImportDefaultSpecifier | ImportNamespaceSpecifier,
+): node is ImportDefaultSpecifier {
+    return node.type === 'ImportDefaultSpecifier';
+}
+
+/** Return false if the merge will produce an invalid result */
+function mergeIsSafe(
+    nodeToKeep: ImportDeclaration,
+    nodeToForget: ImportDeclaration,
+) {
+    if (
+        nodeToKeep.specifiers.some(nodeIsImportNamespaceSpecifier) ||
+        nodeToForget.specifiers.some(nodeIsImportNamespaceSpecifier)
+    ) {
+        // An `import * as Foo` namespace specifier cannot be merged
+        //   with other import expressions.
+        return false;
+    }
+    if (
+        nodeToKeep.specifiers.some(nodeIsImportDefaultSpecifier) &&
+        nodeToForget.specifiers.some(nodeIsImportDefaultSpecifier)
+    ) {
+        // Two `import Foo from` specifiers cannot be merged trivially.
+        // -- Notice: this is *not* import {default as Foo1, default as Foo2} -- that's legal!
+        //
+        // Future work could convert `import Foo1 from 'a'; import Foo2 from 'a';
+        //  into `import {default as Foo1, default as Foo2} from 'a';`
+        // But since this runs the risk of making code longer, this won't be in v1.
+        return false;
+    }
+    return true;
+}
+
+/**
+ * @returns (true to delete node | false to keep node)
+ */
+function mergeNodes(
+    nodeToKeep: ImportDeclaration,
+    nodeToForget: ImportDeclaration,
+) {
+    if (!mergeIsSafe(nodeToKeep, nodeToForget)) {
+        return false;
+    }
+
+    nodeToKeep.specifiers.push(...nodeToForget.specifiers);
+
+    // The line numbers will be all messed up. Is this a problem?
+    nodeToKeep.leadingComments = [
+        ...(nodeToKeep.leadingComments || []),
+        ...(nodeToForget.leadingComments || []),
+    ];
+    nodeToKeep.innerComments = [
+        ...(nodeToKeep.innerComments || []),
+        ...(nodeToForget.innerComments || []),
+    ];
+    nodeToKeep.trailingComments = [
+        ...(nodeToKeep.trailingComments || []),
+        ...(nodeToForget.trailingComments || []),
+    ];
+
+    return true;
+}
+
+/**
+ * Modifies context, deleteContext,
+ * case A: context has no node for an import source, then it's assigned.
+ * case B: context has a node for an import source, then it's merged, and old node is added to deleteContext.
+ */
+function mutateContextAndMerge({
+    context,
+    deleteContext,
```

Comment:
Might as well call this `nodesToDelete`, which is pretty clear, and also the name of the variable used when calling this function.